x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/ExponentialHistogramBlockBuilder.java

```
@@ -0,0 +1,95 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.data;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.Releasables;
+import org.elasticsearch.exponentialhistogram.ExponentialHistogram;
+
+public class ExponentialHistogramBlockBuilder implements Block.Builder {
+
+    private static final int INITIAL_SCRATCH_SIZE = 256;
+
+    private final BytesRefBlock.Builder encodedHistogramsBuilder;
+
+    private final BytesRef tempScratch;
+
+    private boolean closed = false;
+
+    ExponentialHistogramBlockBuilder(int estimatedSize, BlockFactory blockFactory) {
+        this.encodedHistogramsBuilder = blockFactory.newBytesRefBlockBuilder(estimatedSize);
+        this.tempScratch = new BytesRef(new byte[INITIAL_SCRATCH_SIZE], 0, INITIAL_SCRATCH_SIZE);
+    }
+
+    public ExponentialHistogramBlockBuilder append(@Nullable ExponentialHistogram value) {
+        if (value == null) {
+            appendNull();
+        } else {
+            encodedHistogramsBuilder.appendBytesRef(ExponentialHistogramArrayBlock.encode(value, tempScratch));
+        }
+        return this;
+    }
+
+    @Override
+    public ExponentialHistogramBlock build() {
+        return new ExponentialHistogramArrayBlock(encodedHistogramsBuilder.build());
+    }
+
+    @Override
+    public ExponentialHistogramBlockBuilder appendNull() {
+        encodedHistogramsBuilder.appendNull();
+        return this;
+    }
+
+    @Override
+    public ExponentialHistogramBlockBuilder beginPositionEntry() {
+        encodedHistogramsBuilder.beginPositionEntry();
+        return this;
+    }
+
+    @Override
+    public ExponentialHistogramBlockBuilder endPositionEntry() {
+        encodedHistogramsBuilder.endPositionEntry();
```

Can we throw UnsupportedOperationException here - we should not have multi-valued with ExponentialHistogram.

---

The `exponential_histogram` does not support multi-values for technical reasons, as the exponential histograms are stored split across multiple doc values and we can't guarantee the order in case of multivalues. There isn't really a use case justifying the effort required to support this.

For ES|QL, my understanding is that blocks are not only used when loading data, but also for intermediate and end results. So while loading a field of type `exponential_histogram`, can never produces multi-values, why wouldn't we allow users to construct multivalues, e.g. via the [VALUES](https://www.elastic.co/docs/reference/query-languages/esql/functions-operators/aggregation-functions) aggregation?

I don't really see much of a use-case for constructing multi-valued exponential histogram columns this way. But is there that much effort behind supporting it? To me it felt like much more effort not supporting it, because a lot of tests construct multivalues, which would need to be adapted then. But my impression might be wrong here.